File: src/assignment6_gps/src/misc/minarea.cpp
#include "opencv2/highgui.hpp"
#include "opencv2/imgproc.hpp"

#include <iostream>

using namespace cv;
using namespace std;

static void help() {
    cout << "This program demonstrates finding the minimum enclosing box, triangle or circle of a set\n"
         << "of points using functions: minAreaRect() minEnclosingTriangle() minEnclosingCircle().\n"
         << "Random points are generated and then enclosed.\n\n"
         << "Press ESC, 'q' or 'Q' to exit and any other key to regenerate the set of points.\n\n"
         << "Call:\n"
         << "./minarea\n"
         << "Using OpenCV v" << CV_VERSION << "\n" << endl;
}

int main() {
    help();
    Mat img(500, 500, CV_8UC3);
    RNG& rng = theRNG();
    while(true) {
        int i, count = rng.uniform(1, 101);
        vector<Point> points;
        // Generate a random set of points
        for( i = 0; i < count; i++ ) {
            Point pt;
            pt.x = rng.uniform(img.cols/4, img.cols*3/4);
            pt.y = rng.uniform(img.rows/4, img.rows*3/4);
            points.push_back(pt);
        }

        // Find the minimum area enclosing circle
        Point2f center;
        float radius = 0;
        minEnclosingCircle(Mat(points), center, radius);
        img = Scalar::all(0);
        // Draw the points
        for( i = 0; i < count; i++ )
            circle( img, points[i], 3, Scalar(0, 0, 255), FILLED, LINE_AA );
        // Draw the circle
        circle(img, center, cvRound(radius), Scalar(0, 255, 255), 1, LINE_AA);
        imshow( "Enclosing Circle", img );
        char key = (char)waitKey();
        if( key == 27 || key == 'q' || key == 'Q' ) // 'ESC'
            break;
    }
    return 0;
}

File: src/fub_visual_gps/src/misc/minarea.py
import numpy as np

import cv2

img_size = 500


def main():
    key = -1
    while key not in (27, ord('q'), ord('Q')):
        count = np.random.randint(1, 101)
        points = np.random.randint(img_size / 4, img_size * 3 / 4, 2 * count).reshape((count, 2))

        # Find the minimum area enclosing circle
        center, radius = cv2.minEnclosingCircle(points)

        img = np.zeros((img_size, img_size, 3), dtype=np.uint8)

        # Draw the points
        for point in points:
            cv2.circle(img, tuple(point), 3, (0, 0, 255), cv2.FILLED, cv2.LINE_AA)

        # Draw the circle
        cv2.circle(img, tuple(np.round(center).astype(np.int)), int(radius), (0, 255, 255), 1, cv2.LINE_AA)

        cv2.imshow('Enclosing Circle', img)
        key = cv2.waitKey(0) & 0xFF

    cv2.destroyAllWindows()


if __name__ == '__main__':
    main()
